README.md
# Logsprinkler

Broadcasts UDP syslog messages to clients

## Description

logsprinkler is a daemon which listens for syslog messages on UDP port 5515. In
addition it listens for clients, logbarrels, who are sending heartbeats on UDP
port 5514. For any logbarrels that are actively heartbeating it broadcasts any
syslog messages it is receiving on 5515 to those active logbarrels on 5514.

# Build

```
# build individual programs
$ (cd logsprinkler; go build)
$ (cd logbarrel; go build)

# or build a Debian package
$ dpkg-buildpackage -us -uc -b -d
```


debian/changelog
logsprinkler (1.9) stable; urgency=medium

  * use go mod

 -- Jesse Hathaway <jesse@mbuki-mvuki.org>  Thu, 24 Oct 2019 10:50:21 -0500

logsprinkler (1.8) stable; urgency=medium

  * increase UDP read buffer to max UDP size 64KiB

 -- Jesse Hathaway <jesse@mbuki-mvuki.org>  Wed, 23 Oct 2019 16:40:13 -0500

logsprinkler (1.7) stable; urgency=medium

  * rebuild with 1.9
  * add version flag

 -- Jesse Hathaway <jesse@mbuki-mvuki.org>  Fri, 05 Jan 2018 10:19:06 -0600

logsprinkler (1.6) stable; urgency=medium

  * allow connecting to remote logsprinklers

 -- Jesse Hathaway <jesse@mbuki-mvuki.org>  Fri, 10 Nov 2017 11:29:10 -0600

logsprinkler (1.5) stable; urgency=low

  * Remove regex matching to increase performance

 --  Jesse Hathaway <jesse@mbuki-mvuki.org>  Fri, 21 Aug 2015 21:15:40 +0000

logsprinkler (1.4) stable; urgency=low

  * read the daemon man page, simplified the init script

 --  Jesse Hathaway <jesse@mbuki-mvuki.org>  Thu, 18 Jun 2015 18:19:39 +0000

logsprinkler (1.3) stable; urgency=low

  * add ability for logsprinkler to only send logs matching logbarrel regular
    expression

 --  Jesse Hathaway <jesse@mbuki-mvuki.org>  Fri, 29 Apr 2015 16:11:25 +0000

logsprinkler (1.2) stable; urgency=low

  * bump syslogd

 --  Jesse Hathaway <jesse@mbuki-mvuki.org>  Fri, 08 May 2015 12:18:25 +0000

logsprinkler (1.1) stable; urgency=low

  * rename to log sprinkler

 --  Jesse Hathaway <jesse@mbuki-mvuki.org>  Tue, 21 Apr 2015 13:44:43 +0000

logsprinkler (1.0) stable; urgency=low

  * Initial release.

 --  Jesse Hathaway <jesse@mbuki-mvuki.org>  Wed, 08 Apr 2015 19:18:25 +0000


main.go
// vim: noexpandtab tabstop=2 shiftwidth=2:
package main

import (
	"fmt"
	"github.com/jeromer/syslogparser"
	"github.com/wolfeidau/syslogasuarus/syslogd"
	"io"
	"log"
	"net"
)

func main() {
	l, err := net.Listen("tcp", ":5514")
	if err != nil {
		log.Fatal(err)
	}
	defer l.Close()

	log_channel := make(chan string)
	new_client_channel := make(chan chan string)
	go func(new_client_channel chan chan string, log_channel chan string) {
		syslog_channel := make(chan syslogparser.LogParts, 1)
		svr := syslogd.NewServer()
		svr.ListenUDP(":5514")
		svr.Start(syslog_channel)

		var client_channel_list []chan string
		var client_channel chan string
		for {
			select {
			case client_channel = <-new_client_channel:
				client_channel_list = append(client_channel_list, client_channel)
				fmt.Println("New client")
			case logParts := <-syslog_channel:
				log_message := fmt.Sprintf("%s %s %s %s %s\n", logParts["timestamp"], logParts["severity"], logParts["hostname"], logParts["tag"], logParts["content"])
				fmt.Println(log_message)
				for _, client_channel = range client_channel_list {
					client_channel <- log_message
				}
			}
		}
	}(new_client_channel, log_channel)

	for {
		// Wait for a connection.
		conn, err := l.Accept()
		if err != nil {
			log.Fatal(err)
		}
		client_log_channel := make(chan string)
		new_client_channel <- client_log_channel
		// Handle the connection in a new goroutine.
		// The loop then returns to accepting, so that
		// multiple connections may be served concurrently.
		go func(c net.Conn, client_log_channel chan string) {
			// Echo all incoming data.
			for {
				io.WriteString(c, <-client_log_channel)
			}
			// Shut down the connection.
			c.Close()
		}(conn, client_log_channel)
	}
}


syslog-sprinklerd.go
// vim: noexpandtab tabstop=2 shiftwidth=2:

// syslog sprinkler daemon
package main

import (
	"fmt"
	"github.com/jeromer/syslogparser"
	"github.com/wolfeidau/syslogasuarus/syslogd"
	"log"
	"net"
	"time"
)

func handleClientHeartbeats(conn *net.UDPConn, clients chan string) {

  var buf [1024]byte

  for {
    n, addr, err := conn.ReadFromUDP(buf[0:])
    if err != nil {
      log.Print("Error: Received bad UDP packet\n")
    } else if string(buf[0:n]) != "HEARTBEAT" {
      log.Print("Error: Received packet without a heatbeat message", string(buf[0:n]), "\n")
    } else {
      clients <- addr.String()
    }
  }
}

func serveSyslogs(conn *net.UDPConn, clients chan string) {
	syslogChannel := make(chan syslogparser.LogParts, 1)
	syslogServer := syslogd.NewServer()
	syslogServer.ListenUDP(":5514")
	syslogServer.Start(syslogChannel)

	activeClients := make(map[string]time.Time)

	for {
		select {
		case clientAddr := <-clients:
			activeClients[clientAddr] = time.Now()
		case logParts := <-syslogChannel:
			pruneDeadClients(activeClients)

			// Example rsyslog message: 'Apr  3 19:23:40 apply02 nginx-apply: this is the message'
			unformattedTime := logParts["timestamp"].(time.Time)
			formattedTime := unformattedTime.Format(time.Stamp)
			logMessage := fmt.Sprintf("%s %s %s%s %s\n", formattedTime, logParts["hostname"], logParts["tag"], ":", logParts["content"])
			for clientAddr, _ := range activeClients {
				udpAddr, _ := net.ResolveUDPAddr("udp4", clientAddr)
				conn.WriteToUDP([]byte(logMessage), udpAddr)
			}
		}
	}
}

func pruneDeadClients(activeClients map[string]time.Time) {
  for clientAddr, lastHeartbeat := range activeClients {
    now := time.Now()
    lastHeartbeatDuration := now.Sub(lastHeartbeat)
    if lastHeartbeatDuration.Seconds() >= 10 {
      delete(activeClients, clientAddr)
    }
  }
}

func checkError(err error) {
  if err != nil {
    log.Fatalf("Fatal error ", err.Error())
  }
}

func main() {
  service := ":5515"
  udpAddr, err := net.ResolveUDPAddr("udp4", service)
  checkError(err)

  conn, err := net.ListenUDP("udp", udpAddr)
  checkError(err)

  clients := make(chan string)
  go handleClientHeartbeats(conn, clients)
  serveSyslogs(conn, clients)
}


logbarrel/logbarrel.go
// vim: noexpandtab tabstop=2 shiftwidth=2:
// logbarrel: syslog sprinkler client

package main

import (
	"encoding/json"
	"flag"
	"fmt"
	"log"
	"net"
	"os"
	"runtime"
	"runtime/pprof"
	"strings"
	"time"
)

// version set from linker
var version string

type RemoteHeartbeatMessage struct {
	Type       string
	SyslogTags []string
}

func receiveSyslogs(conn net.Conn) {
	readBuffer := make([]byte, 65507)
	for {
		bytesRead, err := conn.Read(readBuffer[0:])
		checkError(err)
		fmt.Print(string(readBuffer[0:bytesRead]))
	}
}

func sendHeartbeats(conn net.Conn, syslogTags []string) {
	var remoteHeartbeatMessage RemoteHeartbeatMessage
	remoteHeartbeatMessage.Type = "heartbeat"
	remoteHeartbeatMessage.SyslogTags = syslogTags
	byteHeartbeatMessage, err := json.Marshal(remoteHeartbeatMessage)
	checkError(err)

	for {
		_, err := conn.Write(byteHeartbeatMessage)
		checkError(err)
		time.Sleep(2 * time.Second)
	}
}

func checkError(err error) {
	if err != nil {
		log.Fatalf("Fatal error: %s", err.Error())
	}
}

func checkErrorFatal(err error) {
	if err != nil {
		log.Fatalf("Fatal error: %s", err.Error())
		os.Exit(1)
	}
}

func main() {
	var syslogTagsString = flag.String("t", "*", "Comma separated list of Syslog tags or '*' to select all")
	var serverIPString = flag.String("s", "localhost", "Hostname or IP of server")
	var showVersion = flag.Bool("v", false, "Show version")
	var cpuprofile = flag.String("cpuprofile", "", "write cpu profile to file")
	var memprofile = flag.String("memprofile", "", "write memory profile to this file")
	var syslogTags []string

	flag.Parse()
	if *showVersion {
		fmt.Printf("logbarrel verision: %s, go runtime version: %s\n", version, runtime.Version())
		os.Exit(0)
	}
	syslogTags = strings.Split(*syslogTagsString, ",")
	serverIP, err := net.ResolveIPAddr("ip", *serverIPString)
	checkError(err)

	if *cpuprofile != "" {
		f, err := os.Create(*cpuprofile)
		checkErrorFatal(err)
		pprof.StartCPUProfile(f)
		defer pprof.StopCPUProfile()
	}

	UDPAddr := net.UDPAddr{IP: serverIP.IP, Port: 5514}
	conn, err := net.DialUDP("udp", nil, &UDPAddr)
	checkError(err)

	go sendHeartbeats(conn, syslogTags)
	receiveSyslogs(conn)

	if *memprofile != "" {
		f, err := os.Create(*memprofile)
		checkErrorFatal(err)
		pprof.Lookup("heap").WriteTo(f, 0)
		f.Close()
	}
}


syslog-sprinkler/syslog-sprinkler.go
// vim: noexpandtab tabstop=2 shiftwidth=2:

// syslog sprinkler daemon
package main

import (
	"fmt"
	"github.com/jeromer/syslogparser"
	"github.com/lollipopman/syslogd"
	"log/syslog"
	"net"
	"os"
	"time"
)

var logger *syslog.Writer

func init() {
		var err error
		logger, err = syslog.Dial("udp", "localhost:514", syslog.LOG_INFO|syslog.LOG_DAEMON, "syslog-sprinkler")
		checkErrorFatal(err)
}

func handleClientHeartbeats(conn *net.UDPConn, clients chan string) {

	var buf [1024]byte

	for {
		n, addr, err := conn.ReadFromUDP(buf[0:])
		if err != nil {
			logger.Info("Error: Received bad UDP packet\n")
		} else if string(buf[0:n]) != "HEARTBEAT" {
			logger.Info("Error: Received packet without a heatbeat message" + string(buf[0:n]) + "\n")
		} else {
			clients <- addr.String()
		}
	}
}

func serveSyslogs(conn *net.UDPConn, clients chan string) {
	var err error
	syslogChannel := make(chan syslogparser.LogParts, 1)
	syslogServer := syslogd.NewServer()
	err = syslogServer.ListenUDP(":5514")
	checkErrorFatal(err)
	syslogServer.Start(syslogChannel)

	activeClients := make(map[string]time.Time)

	for {
		select {
		case clientAddr := <-clients:
			activeClients[clientAddr] = time.Now()
		case logParts := <-syslogChannel:
			pruneDeadClients(activeClients)

			// Example rsyslog message: 'Apr  3 19:23:40 apply02 nginx-apply: this is the message'
			unformattedTime := logParts["timestamp"].(time.Time)
			formattedTime := unformattedTime.Format(time.Stamp)
			logMessage := fmt.Sprintf("%s %s %s%s %s\n", formattedTime, logParts["hostname"], logParts["tag"], ":", logParts["content"])
			for clientAddr, _ := range activeClients {
				udpAddr, _ := net.ResolveUDPAddr("udp4", clientAddr)
				conn.WriteToUDP([]byte(logMessage), udpAddr)
			}
		}
	}
}

func pruneDeadClients(activeClients map[string]time.Time) {
	for clientAddr, lastHeartbeat := range activeClients {
		now := time.Now()
		lastHeartbeatDuration := now.Sub(lastHeartbeat)
		if lastHeartbeatDuration.Seconds() >= 10 {
			delete(activeClients, clientAddr)
		}
	}
}

func checkErrorFatal(err error) {
	if err != nil {
		logger.Crit("Fatal error " + err.Error())
		os.Exit(1)
	}
}

func main() {
	logger.Info("syslog-sprinkler starting up")
	service := ":5515"
	udpAddr, err := net.ResolveUDPAddr("udp4", service)
	checkErrorFatal(err)

	conn, err := net.ListenUDP("udp", udpAddr)
	checkErrorFatal(err)

	clients := make(chan string)
	go handleClientHeartbeats(conn, clients)
	serveSyslogs(conn, clients)
}


logsprinkler/logsprinkler.go
// vim: noexpandtab tabstop=2 shiftwidth=2:
// syslog sprinkler daemon
package main

import (
	"encoding/json"
	"flag"
	"fmt"
	"log/syslog"
	"net"
	"os"
	"os/signal"
	"runtime"
	"runtime/pprof"
	"syscall"
	"time"

	"github.com/jeromer/syslogparser"
	"github.com/lollipopman/syslogd"
)

// version set from linker
var version string

var logger *syslog.Writer

type RemoteHeartbeatMessage struct {
	Type       string
	SyslogTags []string
}

type ClientConfigMessage struct {
	SyslogTags     []string
	NetworkAddress net.UDPAddr
	LastHeartbeat  time.Time
}

func handleClientHeartbeats(conn *net.UDPConn, clients chan ClientConfigMessage) {
	var remoteHeartbeatMessage RemoteHeartbeatMessage
	var clientConfigMessage ClientConfigMessage
	byteHeartbeatMessage := make([]byte, 1024)

	for {
		bytesRead, networkAddress, err := conn.ReadFromUDP(byteHeartbeatMessage)
		if checkError(err) {
			continue
		}
		err = json.Unmarshal(byteHeartbeatMessage[:bytesRead], &remoteHeartbeatMessage)
		if checkError(err) {
			continue
		}
		clientConfigMessage.NetworkAddress = *networkAddress
		clientConfigMessage.LastHeartbeat = time.Now()
		clientConfigMessage.SyslogTags = make([]string, len(remoteHeartbeatMessage.SyslogTags))
		copy(clientConfigMessage.SyslogTags, remoteHeartbeatMessage.SyslogTags)
		clients <- clientConfigMessage
	}
}

func serveSyslogs(conn *net.UDPConn, clients chan ClientConfigMessage, signals chan os.Signal) {
	var err error

	deadClientsTicker := time.NewTicker(time.Second * 10).C
	activeClients := make(map[string]map[string]ClientConfigMessage)
	// logs are about 330 bytes, lets cache up to a 50MBs worth of logs
	syslogChannel := make(chan syslogparser.LogParts, 150000)

	syslogServer := syslogd.NewServer()
	err = syslogServer.ListenUDP(":5515")
	checkErrorFatal(err)
	syslogServer.Start(syslogChannel)

	for {
		select {
		case clientConfigMessage := <-clients:
			addClient(activeClients, clientConfigMessage)
		case <-deadClientsTicker:
			pruneDeadClients(activeClients)
		case logParts := <-syslogChannel:
			sendLogsToClients(conn, activeClients, logParts)
		case signal := <-signals:
			fmt.Println("\nExiting caught signal:", signal)
			return
		}
	}
}

func sendLogsToClients(conn *net.UDPConn, activeClients map[string]map[string]ClientConfigMessage, logParts syslogparser.LogParts) {
	// Example rsyslog message: 'Apr  3 19:23:40 apply02 nginx-apply: this is the message'
	unformattedTime := logParts["timestamp"].(time.Time)
	formattedTime := unformattedTime.Format(time.Stamp)
	logMessage := fmt.Sprintf("%s %s %s%s %s\n", formattedTime, logParts["hostname"], logParts["tag"], ":", logParts["content"])
	syslogTag := fmt.Sprintf("%s", logParts["tag"])

	if clients, ok := activeClients[syslogTag]; ok {
		for _, clientConfigMessage := range clients {
			conn.WriteToUDP([]byte(logMessage), &clientConfigMessage.NetworkAddress)
		}
	}

	// Send all messages to "*" clients
	if clients, ok := activeClients["*"]; ok {
		for _, clientConfigMessage := range clients {
			conn.WriteToUDP([]byte(logMessage), &clientConfigMessage.NetworkAddress)
		}
	}
}

func addClient(activeClients map[string]map[string]ClientConfigMessage, clientConfigMessage ClientConfigMessage) {
	var newClientMap map[string]ClientConfigMessage
	for _, syslogTag := range clientConfigMessage.SyslogTags {
		if clients, ok := activeClients[syslogTag]; ok {
			clients[clientConfigMessage.NetworkAddress.String()] = clientConfigMessage
		} else {
			newClientMap = make(map[string]ClientConfigMessage)
			newClientMap[clientConfigMessage.NetworkAddress.String()] = clientConfigMessage
			activeClients[syslogTag] = newClientMap
		}
	}
}

func pruneDeadClients(activeClients map[string]map[string]ClientConfigMessage) {
	for syslogTag, clients := range activeClients {
		for clientNetworkAddress, clientConfigMessage := range clients {
			lastHeartbeatDuration := time.Now().Sub(clientConfigMessage.LastHeartbeat)
			if lastHeartbeatDuration.Seconds() >= 10 {
				delete(clients, clientNetworkAddress)
			}
		}

		if len(clients) == 0 {
			delete(activeClients, syslogTag)
		} else {
			logger.Info(fmt.Sprintf("Tag: %s, Clients: %d\n", syslogTag, len(clients)))
		}
	}
}

func checkErrorFatal(err error) {
	if err != nil {
		logger.Crit("Fatal error " + err.Error())
		os.Exit(1)
	}
}

func checkError(err error) bool {
	foundError := false
	if err != nil {
		foundError = true
		logger.Warning("Error " + err.Error())
	}
	return foundError
}

func init() {
	var err error
	logger, err = syslog.Dial("udp", "localhost:514", syslog.LOG_INFO|syslog.LOG_DAEMON, "logsprinkler")
	checkErrorFatal(err)
}

func main() {
	var cpuprofile = flag.String("cpuprofile", "", "write cpu profile to file")
	var memprofile = flag.String("memprofile", "", "write memory profile to this file")
	var showVersion = flag.Bool("v", false, "Show version")
	flag.Parse()

	if *showVersion {
		fmt.Printf("logbarrel verision: %s, go runtime version: %s\n", version, runtime.Version())
		os.Exit(0)
	}

	logger.Info("Starting up")

	signals := make(chan os.Signal, 1)
	signal.Notify(signals, syscall.SIGINT, syscall.SIGTERM)

	if *cpuprofile != "" {
		f, err := os.Create(*cpuprofile)
		checkErrorFatal(err)
		pprof.StartCPUProfile(f)
		defer pprof.StopCPUProfile()
	}

	UDPAddr := net.UDPAddr{Port: 5514}
	conn, err := net.ListenUDP("udp", &UDPAddr)
	checkErrorFatal(err)
	clients := make(chan ClientConfigMessage)

	go handleClientHeartbeats(conn, clients)
	serveSyslogs(conn, clients, signals)

	if *memprofile != "" {
		f, err := os.Create(*memprofile)
		checkErrorFatal(err)
		pprof.Lookup("heap").WriteTo(f, 0)
		f.Close()
	}
}
